cmake-modules/add_nlohmann_json.cmake

```
@@ -0,0 +1,30 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+#
+# Defines utility functions to create code coverage targets with gcov.
+# gcov html and xml report.
+#
+
+# Azure Core can expose the Nlohmann Json header by fetching a release version and including the header to the include path.
+# 
+# Azure Core will try to find the json header first. This is for an application that is already using Nlohmann header.
+# When it is not found it, Azure Core will featch it.
+
+# Storage requires 3.8 version for using `contains` feature
+find_package(nlohmann_json 3.8.0 CONFIG)
```

What happens here, if the application already uses a `nlohmnn_json` but with a different version (higher or lower)? Does `find_package` still find it?

I am not sure about this so this is an open question, but should we use `EXACT` match?

https://cmake.org/cmake/help/v3.0/command/find_package.html
> When the [version] argument is given Config mode will only find a version of the package that claims compatibility with the requested version (format is major[.minor[.patch[.tweak]]]). If the EXACT option is given only a version of the package claiming an exact match of the requested version may be found.

---

Valid question. If the version is lower, CMake would consider it as not found, so it would fetch the version we need.
Then, it might collapse with customer's version, and fail on compile time.

If the version is higher, CMake consider it as FOUND and use that higher version.

So, the only problem here is when the version is lower.
I tried to use an old version (3.2) but Storage was already using 3.8 and got an error because `contains` API was missing, so I updated it.

This is one of the things we want to define for the final approach.

---

nit: In that case, can we use exact for now, to require precisely the version we need. I don't know if we want to have "silent" upgrades to higher versions because I am not sure if `nlohmnn_json` always guarantees backwards compatibility.

```suggestion
find_package(nlohmann_json 3.8.0 EXACT)
```

---

why not just `find_package(nlohmann_json)`?

How do you know version below 3.8.0 won't work for us?

---

> How do you know version below 3.8.0 won't work for us?

See @vhvb1989's comment above. Looks like you are using APIs and features that were added in newer versions, like `contains`:
https://github.com/nlohmann/json/releases/tag/v3.6.0
> introduces a contains function to check if a key is present in an object,
https://github.com/Azure/azure-sdk-for-cpp/blob/71c5cca9d59dd1ab6d61be917090b19a5c4175af/sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp#L1259

> I tried to use an old version (3.2) but Storage was already using 3.8 and got an error because `contains` API was missing, so I updated it.

---

Not to mention, that's the version of the dependency that's currently checked-in:
https://github.com/Azure/azure-sdk-for-cpp/blob/71c5cca9d59dd1ab6d61be917090b19a5c4175af/sdk/storage/azure-storage-common/inc/azure/storage/common/json.hpp#L4

---

fine, but shouldn't be 3.6.0 rather than 3.8.0 then?

Also, I believe `EXACT` is not a good idea.

---

ok, will use EXACT + 3.6.0

---

> fine, but shouldn't be 3.6.0 rather than 3.8.0 then?

If we are open to changing it, it should be the most recent version.

---

the most recent version is 3.9.1 https://github.com/nlohmann/json/releases